src/charm.py

```diff
@@ -79,29 +117,38 @@ def _on_start(self, event: ops.charm.StartEvent) -> None:
                 return
 
         try:
+            logger.debug("opening tcp port")
             self._open_port_tcp(self._port)
         except subprocess.CalledProcessError:
             self.unit.status = BlockedStatus("failed to open TCP port for MongoDB")
             return
 
-        if not self._mongo.is_ready():
+        # check if this units deployment of MongoDB is ready
+        if not self._mongo.is_ready(all_replicas=False):
+            logger.debug("mongoDB not ready, defering on start event")
             self.unit.status = WaitingStatus("waiting for MongoDB to start")
             event.defer()
             return
 
-        # initialize replica set
-        self.unit.status = MaintenanceStatus("initialising MongoDB replicaset")
-        logger.debug("initialising replica set")
-        try:
-            logger.debug("initialising replica set for the following IPs %s", self._unit_ips)
-            self._mongo.initialize_replica_set(self._unit_ips)
-            self._peers.data[self.app]["replica_set_hosts"] = json.dumps(self._unit_ips)
-        except (ConnectionFailure, ConfigurationError) as e:
-            logger.error("error initialising replica sets in _on_start: error: %s", str(e))
-            self.unit.status = BlockedStatus("failed to initialise replicaset")
+        # mongod is now active
+        self.unit.status = ActiveStatus()
+
+        # only leader should initialise the replica set
```

Comment:
We can probably let the `_mongodb_relation_joined()` event handler to manage the replicaset intialisation and let `_on_start()` manage the unit initialisation. I'm seeing this block of code is in both, so we might as well let just one event handler to do the task. I'm not sure if that works for MongoDB, though, this is just a suggestion.

Reply:
Ah, that makes sense.